File: simulated_annealing/sa.py
import time 
import math
import numpy as np
import argparse
import matplotlib as mpl
import random as rand
from matplotlib import pyplot as plt
from mpl_toolkits.mplot3d import Axes3D

class SatInt:
    def __init__(self, val, lo, hi):
        self.real, self.lo, self.hi = val, lo, hi

    def __add__(self, other):
        return min(self.real + other.real, self.hi)

    def __sub__(self, other):
        return max(self.real - other.real, self.lo)

class SimulatedAnnealing:

    def __init__(self) -> None:
        
        # Alpha value for decrementing temperature
        # Linear annealing schedule
        # self.alpha = 0.001
        # Geometric annealing schedule
        self.alpha = 0.95

        # Initial temperature
        self.t = 0.95

        # Initial starting point 
        self.start = (rand.randint(-100, 100), rand.randint(-100, 100))

        # Global min to date
        # [x1, x2, z]
        self.g_min = [0,0,0]

        # All points visited
        self.path = []

        # Track Iterations
        self.iter = 1

        # Hard limit on max iterations
        # Linear annealing schedule
        # self.max_iter = round((1/(self.alpha)))
        
        # Geometric annealing schedule
        self.max_iter = 10000

        print(f"Initial: Alpha={self.alpha}, t={self.t}, start={self.start}, max_iter={self.max_iter}")

    def __objective_func(self, x1, x2):
        z = -np.cos(x1)*np.cos(x2)*np.exp(-np.power((x1 - math.pi), 2) - np.power((x2 - math.pi), 2))
        return z

    def __acceptance_probability(self):
        return rand.random()

    def __probability_func(self, delta):
        return np.exp(-delta/self.t)
    
    def __neighborhood_func(self, x1, x2):

        x1_new = SatInt(x1, -100, 100)
        x2_new = SatInt(x2, -100, 100)

        x1_new = x1_new + rand.random()*rand.randint(-1, 1)
        x2_new = x1_new + rand.random()*rand.randint(-1, 1)

        return x1_new, x2_new

    def search(self, debug=False):

        # Start
        z0 = self.__objective_func(self.start[0], self.start[1])
        p0 = [self.start[0], self.start[1], z0]
        self.g_min = p0
        self.path.append(p0)

        # While temperature is above 0 and max_iter is not exceeded
        while(self.t > 0 and self.iter < self.max_iter):

            # Use to track succesful iteration
            iterate = False

            # Get current step
            x1 = self.path[-1:][0][0]
            x2 = self.path[-1:][0][1]
            z = self.path[-1:][0][2]

            # Generate step
            x1n, x2n = self.__neighborhood_func(x1, x2)

            # Evaluate step
            zn = self.__objective_func(x1n, x2n)

            # See if step is better
            delta = abs(z) - abs(zn)
            if (delta < 0):

                # Move to next state
                pn = [x1n, x2n, zn]
                self.g_min = pn
                self.path.append(pn)

                # Iterate
                iterate = True
            else:
                # Calculate probability 
                ft = self.__probability_func(delta)
                accept = self.__acceptance_probability()

                if accept < ft:
                    # Move to next state
                    pn = [x1n, x2n, zn]
                    self.path.append(pn)

                    # store min only if its better
                    if abs(zn) > abs(self.g_min[2]):
                        self.g_min = pn

                    # Iterate
                    iterate = True

            if iterate:

                # update temperature & iterations
                # Linear annealing schedule
                # self.t = self.t - self.alpha
                
                # Geometric annealing schedule
                self.t = self.t*self.alpha

                self.iter = self.iter + 1

            if debug:
                print(f"\nIteration[{self.iter}], t={self.t}, global_min={self.g_min}")
                print(f"Step [{self.path[-1:][0]}]")
        
        print(f"END: start={self.start}, min={self.g_min} path_len={len(self.path)}")

File: simulated_annealing/rust/src/sa.rs
use rand::Rng;

pub struct SimulatedAnnealing1D {

    // Alpha value for decrementing temperature
    pub alpha: f32,

    // Initial temperature value
    pub temperature: f32,

    // Starting point - 1D
    pub start_x: f32,

    // Maximun number of iterations
    pub max_iterations: i32,

    // Track search path - x,y values
    path: Vec<(f32, f32)>,

    // Track total number of iterations
    iterations: i32,

    // Random number generator
    rng: rand::rngs::ThreadRng,

    // Constraint x to a specific domain
    search_domain_x: (f32, f32),

    // Track best solution [x, y]
    best_solution: (f32, f32),

}

impl SimulatedAnnealing1D {
    
    // Constructor
    pub fn new(a: f32, it: f32, sx: f32, mi: i32) -> SimulatedAnnealing1D {
        SimulatedAnnealing1D {
            alpha: a,
            temperature: it, // set initial temperature
            start_x: sx,
            max_iterations: mi,
            path: Vec::new(),
            iterations: 0,
            rng: rand::thread_rng(),
            search_domain_x: (0.0, 1.2),
            best_solution: (0.0, 0.0),
        }
    }

    // objecttive function
    fn objective_function(&self, x: f32) -> f32 {
        // 1.6 * x * sin(18.0*x)
        return 1.6 * x * f32::sin(18.0*x);
    }

    // Acceptance probability
    fn acceptance_probability(&mut self) -> f32 {

        // Generate a random number between 0 and 1
        return self.rng.gen_range(0.0..1.0);
    }
    
    // Probability function
    fn probability_function(&self, delta: f32) -> f32 {
        return (-delta / self.temperature).exp();
    }

    // Neighborhood function
    fn neighborhood_function(&mut self, x: f32) -> f32 {

        // Increment random in steps of up to 0.1
        let mut new_x: f32 = x + self.rng.gen_range(-0.25..0.25);

        // Constrain x to search domain
        new_x = new_x.max(self.search_domain_x.0);
        new_x = new_x.min(self.search_domain_x.1);

        return new_x;
    }

    // Run search
    pub fn search(&mut self) {

        // Initialize start
        let mut current_x: f32 = self.start_x;
        let mut current_solution: f32 = self.objective_function(current_x);

        self.best_solution = (current_x, current_solution);
        self.path.push((current_x, current_solution));

        // Loop until max iterations
        while self.temperature > 0.0 && self.iterations < self.max_iterations {

            // Use to determine if to iterate to the next temperature
            let mut interate: bool = false;

            // Get current x & solution
            current_x = self.path.last().unwrap().0;
            current_solution = self.path.last().unwrap().1;

            // Generate new x
            let new_x: f32 = self.neighborhood_function(current_x);

            // Determine if new x is better than the best so far x
            let current_new_solution: f32 = self.objective_function(new_x);

            // Here we are trying to find the global minimun
            // If new solution is better than the best so far
            if current_new_solution < current_solution {
                
                // Update path
                self.path.push((current_x, current_new_solution));
                interate = true;

            // If not use probability_function to determine if to go there anyway
            } else {

                let delta: f32 = current_new_solution.abs() - current_solution.abs();
                let probability: f32 = self.probability_function(delta);
                let accept_p: f32 = self.acceptance_probability();

                // If accept_p is less than probability then go there anyway
                if accept_p < probability {

                    // Update path
                    self.path.push((current_x, current_new_solution));
                    interate = true;
                }
            }

            // Check if current solution is best so far
            if current_new_solution < self.best_solution.1 {
                self.best_solution = (new_x, current_new_solution);
            }

            // Check if temperature should decrease
            if interate {

                // Using geometric cooling schedule
                self.temperature = self.temperature * self.alpha;

                // Update iterations
                self.iterations += 1;
            }
        }
    }

    pub fn get_stats(&self) -> (f32, f32, i32) {
        return (self.best_solution.0, self.best_solution.1, self.iterations);
    }

}